[SPARK-21882][CORE] OutputMetrics doesn't count written bytes correctly in the saveAsHadoopDataset function

awarrior opened the pull request:
[SPARK-21882](https://issues.apache.org/jira/browse/SPARK-21882)

## What changes were proposed in this pull request?

The first job called from saveAsHadoopDataset, running in each executor, does not calculate the writtenBytes of OutputMetrics correctly (writtenBytes is 0). The reason is that we did not initialize the callback function called to find bytes written in the right way. As usual, statisticsTable which records statistics in a FileSystem must be initialized at the beginning (this will be triggered when open SparkHadoopWriter). One solution for this issue is to switch the initialization order of HadoopOutputMetrics and SparkHadoopWriter.

## How was this patch tested?

Existing tests

awarrior committed:
Update SparkHadoopWriter.scala

spark-21882

AmplabJenkins commented:
Can one of the admins verify this patch?

jiangxb1987 commented:
ok to test

jiangxb1987 commented:
This should be a valid fix. Could you add a test case for this?

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #81579 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81579/testReport)** for PR 19118 at commit [`0f0c3b1`](https://github.com/apache/spark/commit/0f0c3b1c91b4f06c7e48874b8f6329c5c1c1b3ce).

SparkQA commented:
**[Test build #81579 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81579/testReport)** for PR 19118 at commit [`0f0c3b1`](https://github.com/apache/spark/commit/0f0c3b1c91b4f06c7e48874b8f6329c5c1c1b3ce).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81579/
Test FAILed.

awarrior commented:
@jiangxb1987 ok. I add one later.

awarrior commented:
I met a trouble when I write a test case. It seems that this issue won't be triggered in only one node. I found that Driver node do createPathFromString so that there is no problem.

>   java.lang.Thread.State: RUNNABLE
	  at org.apache.hadoop.fs.FileSystem.getStatistics(FileSystem.java:3271)
	  - locked <0x1211> (a java.lang.Class)
	  at org.apache.hadoop.fs.FileSystem.initialize(FileSystem.java:202)
	  at org.apache.hadoop.fs.RawLocalFileSystem.initialize(RawLocalFileSystem.java:92)
	  at org.apache.hadoop.fs.LocalFileSystem.initialize(LocalFileSystem.java:47)
	  at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2598)
	  at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:91)
	  at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2632)
	  at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2614)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:370)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:169)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:354)
	  at org.apache.hadoop.fs.Path.getFileSystem(Path.java:296)
	  at org.apache.spark.internal.io.SparkHadoopWriterUtils$.createPathFromString(SparkHadoopWriterUtils.scala:55)

Does anyone know how to test in this case?

jiangxb1987 commented:
You don't have to use `saveAsHadoopFile`, just call `saveAsHadoopDataset` directly?

viirya commented on core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala:
```diff
@@ -112,11 +112,12 @@ object SparkHadoopWriter extends Logging {
     config.initWriter(taskContext, sparkPartitionId)
     var recordsWritten = 0L
+    
+    // Initialize callback function after the writer.
```
The comment is better to explain why we must initialize the callback here. E.g., `We must initialize the callback for calculating bytes written after the statistic table is initialized in FileSystem`.

awarrior replied:
ok

awarrior commented:
@jiangxb1987 well, I passed that part above but met other initialization chances before runJob. They are in the write function of SparkHadoopWriter.

>     
    // Assert the output format/key/value class is set in JobConf.
    config.assertConf(jobContext, rdd.conf)                //// <= chance

    val committer = config.createCommitter(stageId)
    committer.setupJob(jobContext)                //// <= chance

    // Try to write all RDD partitions as a Hadoop OutputFormat.
    try {
      val ret = sparkContext.runJob(rdd, (context: TaskContext, iter: Iterator[(K, V)]) => {
        executeTask(
          context = context,
          config = config,
          jobTrackerId = jobTrackerId,
          sparkStageId = context.stageId,
          sparkPartitionId = context.partitionId,
          sparkAttemptNumber = context.attemptNumber,
          committer = committer,
          iterator = iter)
      })

One trace list:

> java.lang.Thread.State: RUNNABLE
	  at org.apache.hadoop.fs.FileSystem.getStatistics(FileSystem.java:3270)
	  - locked <0x126a> (a java.lang.Class)
	  at org.apache.hadoop.fs.FileSystem.initialize(FileSystem.java:202)
	  at org.apache.hadoop.fs.RawLocalFileSystem.initialize(RawLocalFileSystem.java:92)
	  at org.apache.hadoop.fs.LocalFileSystem.initialize(LocalFileSystem.java:47)
	  at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2598)
	  at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:91)
	  at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2632)
	  at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2614)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:370)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:169)
	  at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:354)
	  at org.apache.hadoop.fs.Path.getFileSystem(Path.java:296)
	  at org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter.<init>(FileOutputCommitter.java:91)
	  at org.apache.hadoop.mapred.FileOutputCommitter.getWrapped(FileOutputCommitter.java:65)
	  at org.apache.hadoop.mapred.FileOutputCommitter.setupJob(FileOutputCommitter.java:131)
	  at org.apache.hadoop.mapred.OutputCommitter.setupJob(OutputCommitter.java:233)
	  at org.apache.spark.internal.io.HadoopMapReduceCommitProtocol.setupJob(HadoopMapReduceCommitProtocol.scala:125)
	  at org.apache.spark.internal.io.SparkHadoopWriter$.write(SparkHadoopWriter.scala:74)

awarrior committed:
Update SparkHadoopWriter.scala

modify comment to explain why we switch the initialization order of HadoopOutputMetrics and SparkHadoopWriter

SparkQA commented:
**[Test build #81662 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81662/testReport)** for PR 19118 at commit [`2c4f2ca`](https://github.com/apache/spark/commit/2c4f2ca7f92916114d090208091ba718da5621c6).

SparkQA commented:
**[Test build #81662 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81662/testReport)** for PR 19118 at commit [`2c4f2ca`](https://github.com/apache/spark/commit/2c4f2ca7f92916114d090208091ba718da5621c6).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81662/
Test FAILed.

jiangxb1987 commented:
ping @awarrior Could you please fix the test failure?

awarrior commented:
@jiangxb1987 well, the test case is hard to construct if we just run app in local like comments above. Any ideas to crack?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

mridulm commented:
The change looks good to me, pending test case; thanks for looking into this @awarrior

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

jiangxb1987 commented:
retest this please

AmplabJenkins commented:
Can one of the admins verify this patch?

jiangxb1987 commented:
retest this please

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #4793 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4793/testReport)** for PR 19118 at commit [`2c4f2ca`](https://github.com/apache/spark/commit/2c4f2ca7f92916114d090208091ba718da5621c6).

SparkQA commented:
**[Test build #4793 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4793/testReport)** for PR 19118 at commit [`2c4f2ca`](https://github.com/apache/spark/commit/2c4f2ca7f92916114d090208091ba718da5621c6).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen reviewed (changes requested):
@awarrior if you'll fix the style issue (line too long I think) we can probably merge this

gatorsmile commented:
@Ngone51 Could you please take this over?

Ngone51 commented:
sure @gatorsmile

srowen commented:
@Ngone51 did you open another PR? I can try if not.

Ngone51 commented:
Sorry, haven't. Please try it, thanks. @srowen

srowen commented:
Moved to https://github.com/apache/spark/pull/24863
